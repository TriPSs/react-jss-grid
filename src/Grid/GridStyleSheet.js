import hasOwnProperty from 'has-own-property'
import generateBreakpoints from '../utils/breakpoints'

export const GUTTERS = [0, 8, 16, 24, 40]
export const GRID_SIZES = [true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

export const generateGrid = (globalStyles, theme, breakpoint) => {
  // For the auto layouting
  const styles = {
    [`grid-${breakpoint}`]: {
      flexBasis: 0,
      flexGrow : 1,
      maxWidth : '100%',
    },
  }

  GRID_SIZES.forEach(size => {
    if (typeof size === 'boolean') {
      // Skip the first one as handle above.
      return
    }

    // Only keep 6 significant numbers.
    const width = `${Math.round(size / 12 * 10e6) / 10e4}%`

    /* eslint-disable max-len */
    // Close to the bootstrap implementation:
    // https://github.com/twbs/bootstrap/blob/b0508a975d711d6b24c01f57dd5445c22699fac4/scss/mixins/_grid.scss#L69
    /* eslint-enable max-len */
    styles[`grid-${breakpoint}-${size}`] = {
      flexBasis: width,
      maxWidth : width,
    }
  })

  // No need for a media query for the first size.
  if (breakpoint === 'xs') {
    Object.assign(globalStyles, styles)
  } else {
    globalStyles[theme.breakpoints.up(breakpoint)] = styles
  }
}

export const generateGutter = (theme, breakpoint) => {
  const styles = {}

  GUTTERS.forEach((spacing, index) => {
    if (index === 0) {
      // Skip the default style.
      return
    }

    styles[`spacing-${breakpoint}-${spacing}`] = {
      margin         : -spacing / 2,
      width          : `calc(100% + ${spacing}px)`,
      '& > $typeItem': {
        padding: spacing / 2,
      },
    }
  })

  return styles
}

// Default CSS values
// flex: '0 1 auto',
// flexDirection: 'row',
// alignItems: 'flex-start',
// flexWrap: 'nowrap',
// justifyContent: 'flex-start',
export default theme => {
  const breakpoints = hasOwnProperty(theme, 'breakpoints') ? theme.breakpoints : generateBreakpoints()

  return {
    typeContainer: {
      boxSizing: 'border-box',
      display  : 'flex',
      flexWrap : 'wrap',
      width    : '100%',
    },

    typeItem: {
      boxSizing: 'border-box',
      flex     : '0 0 auto',
      margin   : '0', // For instance, it's useful when used with a `figure` element.
    },

    zeroMinWidth: {
      minWidth: 0,
    },

    flex: {
      display: 'flex',
    },

    'direction-xs-column': {
      flexDirection: 'column',
    },

    'direction-xs-column-reverse': {
      flexDirection: 'column-reverse',
    },

    'direction-xs-row-reverse': {
      flexDirection: 'row-reverse',
    },

    'wrap-xs-nowrap': {
      flexWrap: 'nowrap',
    },

    'wrap-xs-wrap-reverse': {
      flexWrap: 'wrap-reverse',
    },

    'align-items-xs-center': {
      alignItems: 'center',
    },

    'align-items-xs-flex-start': {
      alignItems: 'flex-start',
    },

    'align-items-xs-flex-end': {
      alignItems: 'flex-end',
    },

    'align-items-xs-baseline': {
      alignItems: 'baseline',
    },

    'align-content-xs-center': {
      alignContent: 'center',
    },

    'align-content-xs-flex-start': {
      alignContent: 'flex-start',
    },

    'align-content-xs-flex-end': {
      alignContent: 'flex-end',
    },

    'align-content-xs-space-between': {
      alignContent: 'space-between',
    },

    'align-content-xs-space-around': {
      alignContent: 'space-around',
    },

    'justify-xs-center': {
      justifyContent: 'center',
    },

    'justify-xs-flex-end': {
      justifyContent: 'flex-end',
    },

    'justify-xs-space-between': {
      justifyContent: 'space-between',
    },

    'justify-xs-space-around': {
      justifyContent: 'space-around',
    },

    ...generateGutter(theme, 'xs'),

    ...breakpoints.keys.reduce((accumulator, key) => {
      // Use side effect over immutability for better performance.
      generateGrid(accumulator, { ...theme, breakpoints }, key)

      return accumulator
    }, {}),
  }
}
