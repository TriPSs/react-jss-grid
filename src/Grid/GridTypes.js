// @flow
import type { Element } from 'react'

export type GridSizes = boolean | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12

export type DefaultProps = {
  classes: Object,
}

export type Props = {
  /**
   * The content of the component.
   */
  children?: Element<*>,
  /**
   * Useful to extend the style applied to components.
   */
  classes?: Object,
  /**
   * @ignore
   */
  className?: string,
  /**
   * The component used for the root node.
   * Either a string to use a DOM element or a component.
   */
  component?: string | Function,
  /**
   * If `true`, the component will have the flex *container* behavior.
   * You should be wrapping *items* with a *container*.
   */
  container?: boolean,
  /**
   * It true, the component will have the flex *item* behavior.
   * You should be wrapping *items* with a *container*.
   */
  item?: boolean,
  /**
   * Defines the `align-items` style property.
   * It's applied for all screen sizes.
   */
  align?: 'flex-start' | 'center' | 'flex-end' | 'stretch' | 'baseline',
  /**
   * Defines the `flex-direction` style property.
   * It is applied for all screen sizes.
   */
  direction?: 'row' | 'row-reverse' | 'column' | 'column-reverse',
  /**
   * Defines the space between the type `item` component.
   * It can only be used on a type `container` component.
   */
  spacing?: 0 | 8 | 16 | 24 | 40,
  /**
   * Defines the `justify-content` style property.
   * It is applied for all screen sizes.
   */
  justify?: 'flex-start' | 'center' | 'flex-end' | 'space-between' | 'space-around',
  /**
   * Defines the `flex-wrap` style property.
   * It's applied for all screen sizes.
   */
  wrap?: 'nowrap' | 'wrap' | 'wrap-reverse',
  /**
   * Defines the number of grids the component is going to use.
   * It's applied for all the screen sizes with the lowest priority.
   */
  xs?: GridSizes,
  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `sm` breakpoint and wider screens if not overridden.
   */
  sm?: GridSizes,
  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `md` breakpoint and wider screens if not overridden.
   */
  md?: GridSizes,
  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `lg` breakpoint and wider screens if not overridden.
   */
  lg?: GridSizes,
  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `xl` breakpoint and wider screens.
   */
  xl?: GridSizes,
}

export type AllProps = DefaultProps & Props
