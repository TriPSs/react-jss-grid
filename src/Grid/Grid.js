import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import withStyles from 'react-jss'

import GridStyleSheet, { GUTTERS } from './GridStyleSheet'

function Grid(props) {
  const {
    alignContent,
    alignItems,
    classes,
    className: classNameProp,
    component: Component,
    container,
    direction,
    hidden,
    flex,
    item,
    justify,
    lg,
    md,
    zeroMinWidth,
    sm,
    spacing,
    wrap,
    xl,
    xs,
    theme,
    ...other
  } = props

  const className = classNames(
    {
      [classes.typeContainer]                              : container,
      [classes.typeItem]                                   : item,
      [classes.zeroMinWidth]                               : zeroMinWidth,
      [classes.flex]                                       : flex,
      [classes[`spacing-xs-${String(spacing)}`]]           : container && spacing !== 0,
      [classes[`direction-xs-${String(direction)}`]]       : direction !== Grid.defaultProps.direction,
      [classes[`wrap-xs-${String(wrap)}`]]                 : wrap !== Grid.defaultProps.wrap,
      [classes[`align-items-xs-${String(alignItems)}`]]    : alignItems !== Grid.defaultProps.alignItems,
      [classes[`align-content-xs-${String(alignContent)}`]]: alignContent !== Grid.defaultProps.alignContent,
      [classes[`justify-xs-${String(justify)}`]]           : justify !== Grid.defaultProps.justify,
      [classes['grid-xs']]                                 : xs === true,
      [classes[`grid-xs-${String(xs)}`]]                   : xs && xs !== true,
      [classes['grid-sm']]                                 : sm === true,
      [classes[`grid-sm-${String(sm)}`]]                   : sm && sm !== true,
      [classes['grid-md']]                                 : md === true,
      [classes[`grid-md-${String(md)}`]]                   : md && md !== true,
      [classes['grid-lg']]                                 : lg === true,
      [classes[`grid-lg-${String(lg)}`]]                   : lg && lg !== true,
      [classes['grid-xl']]                                 : xl === true,
      [classes[`grid-xl-${String(xl)}`]]                   : xl && xl !== true,
    },
    classNameProp,
  )

  const gridProps = { className, ...other }

  return <Component {...gridProps} />
}

Grid.propTypes = {
  /**
   * Defines the `align-content` style property.
   * It's applied for all screen sizes.
   */
  alignContent: PropTypes.oneOf([
    'stretch',
    'center',
    'flex-start',
    'flex-end',
    'space-between',
    'space-around',
  ]),
  /**
   * Defines the `align-items` style property.
   * It's applied for all screen sizes.
   */
  alignItems  : PropTypes.oneOf(['flex-start', 'center', 'flex-end', 'stretch', 'baseline']),
  /**
   * The content of the component.
   */
  children    : PropTypes.node,
  /**
   * Useful to extend the style applied to components.
   */
  classes     : PropTypes.object.isRequired,
  /**
   * @ignore
   */
  className   : PropTypes.string,

  /**
   * The component used for the root node.
   * Either a string to use a DOM element or a component.
   */
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /**
   * If `true`, the component will have the flex *container* behavior.
   * You should be wrapping *items* with a *container*.
   */
  container: PropTypes.bool,

  /**
   * Defines the `flex-direction` style property.
   * It is applied for all screen sizes.
   */
  direction: PropTypes.oneOf(['row', 'row-reverse', 'column', 'column-reverse']),

  /**
   * If provided, will wrap with [Hidden](/api/hidden) component and given properties.
   */
  hidden: PropTypes.object,

  /**
   * If `true`, the component will have display flex
   */
  flex: PropTypes.bool,

  /**
   * If `true`, the component will have the flex *item* behavior.
   * You should be wrapping *items* with a *container*.
   */
  item: PropTypes.bool,

  /**
   * Defines the `justify-content` style property.
   * It is applied for all screen sizes.
   */
  justify: PropTypes.oneOf(['flex-start', 'center', 'flex-end', 'space-between', 'space-around']),

  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `lg` breakpoint and wider screens if not overridden.
   */
  lg: PropTypes.oneOf([true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),

  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `md` breakpoint and wider screens if not overridden.
   */
  md: PropTypes.oneOf([true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),

  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `sm` breakpoint and wider screens if not overridden.
   */
  sm: PropTypes.oneOf([true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),

  /**
   * Defines the space between the type `item` component.
   * It can only be used on a type `container` component.
   */
  spacing: PropTypes.oneOf(GUTTERS),

  /**
   * Defines the `flex-wrap` style property.
   * It's applied for all screen sizes.
   */
  wrap: PropTypes.oneOf(['nowrap', 'wrap', 'wrap-reverse']),

  /**
   * Defines the number of grids the component is going to use.
   * It's applied for the `xl` breakpoint and wider screens.
   */
  xl: PropTypes.oneOf([true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),

  /**
   * Defines the number of grids the component is going to use.
   * It's applied for all the screen sizes with the lowest priority.
   */
  xs: PropTypes.oneOf([true, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),

  /**
   * If `true`, it sets `min-width: 0` on the item.
   * Refer to the limitations section of the documentation to better understand the use case.
   */
  zeroMinWidth: PropTypes.bool,
}

Grid.defaultProps = {
  alignContent: 'stretch',
  alignItems  : 'stretch',
  component   : 'div',
  container   : false,
  flex        : false,
  direction   : 'row',
  item        : false,
  justify     : 'flex-start',
  zeroMinWidth: false,
  spacing     : 16,
  wrap        : 'wrap',
}

export default withStyles(GridStyleSheet, { name: 'JssGrid' })(Grid)
