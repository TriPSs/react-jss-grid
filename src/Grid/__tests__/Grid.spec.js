// @flow

import React from 'react'
import { ThemeProvider } from 'react-jss'
import renderer from 'react-test-renderer';
import Grid from '../Grid'

describe('(Component) <Grid />', () => {

  const render = children => renderer.create(
    <ThemeProvider theme={{}}>
      {children}
    </ThemeProvider>
  )

  it('should render', () => {
    const component = render(<Grid className="woofGrid" />).toJSON()

    expect(component).toMatchSnapshot();
    expect(component.type).toBe('div')
    expect(component.props.className).toBe('woofGrid')
  })


  describe('prop: container', () => {
    it('should apply the container class', () => {
      const component = render(<Grid container />).toJSON()

      expect(component).toMatchSnapshot();
      expect(component.props.className).toContain('typeContainer')
    })
  })

  describe('prop: item', () => {
    it('should apply the item class', () => {
      const component = render(<Grid item />).toJSON()

      expect(component).toMatchSnapshot();
      expect(component.props.className).toContain('typeItem')
    })
  })

  describe('prop: component', () => {
    it('should use the default', () => {
      const component = render(<Grid />).toJSON()

      expect(component.type).toBe('div')
    })

    it('should change the component', () => {
      const component = render(<Grid component={'span'} />).toJSON()

      expect(component.type).toBe('span')
    })
  })

  describe('prop: xs', () => {
    it('should apply the flex-grow class', () => {
      const component = render(<Grid item xs />).toJSON()

      expect(component.props.className).toContain('grid-xs')
    })

    it('should apply the flex size class', () => {
      const component = render(<Grid item xs={3} />).toJSON()

      expect(component.props.className).toContain('grid-xs-3')
    })
  })

  describe('prop: spacing', () => {
    it('should have a default spacing', () => {
      const component = render(<Grid container />).toJSON()

      expect(component.props.className).toContain('spacing-xs-16')
    })
  })

  describe('prop: other', () => {
    it('should spread the other properties to the root element', () => {
      const handleClick = () => {}

      const component = render(<Grid component={'span'} onClick={handleClick} />).toJSON()
      expect(component.props.onClick).toEqual(handleClick)
    })
  })
})
