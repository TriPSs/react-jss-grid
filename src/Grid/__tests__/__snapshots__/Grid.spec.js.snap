// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) <Grid /> prop: container should apply the container class 1`] = `
<div
  className="typeContainer-0-83 spacing-xs-16-0-98"
/>
`;

exports[`(Component) <Grid /> prop: item should apply the item class 1`] = `
<div
  className="typeItem-0-167"
/>
`;

exports[`(Component) <Grid /> should render 1`] = `
<div
  className="woofGrid"
/>
`;
